includes/subscriptions/class-wc-payments-invoice-service.php

```diff
@@ -227,7 +227,7 @@ public static function get_pending_invoice_id( WC_Subscription $subscription ) :
 	 * @param WC_Subscription $subscription The subscription.
 	 * @param string          $invoice_id   The invoice ID.
 	 */
-	private function set_pending_invoice_id( WC_Subscription $subscription, string $invoice_id ) {
```

Comment:
Removing these types from the function signature isn't ideal however they were causing a number of unit testing related issues.

Namely, this [`test_card_is_saved_when_updating_subscription_using_saved_payment_method()`](https://github.com/Automattic/woocommerce-payments/blob/add/subscription_change_method/tests/unit/test-class-wc-payment-gateway-wcpay-subscriptions-process-payment.php#L371) unit test  function creates an order object and passes it to `$this->mock_wcpay_gateway->process_payment( $order->get_id() );` however that order is suppose to behave like a subscription (note the [`WC_Subscriptions::set_wcs_is_subscription()`](https://github.com/Automattic/woocommerce-payments/blob/add/subscription_change_method/tests/unit/test-class-wc-payment-gateway-wcpay-subscriptions-process-payment.php#L381) call) to test changing the payment method. 

That process payment call eventually leads to `maybe_attempt_payment_for_subscription()` which then tries to get the pending invoice by calling `get_pending_invoice_id()` because the object is an order an not a subscription, the type error is thrown.

I tried changing that unit test to instead use the mock `WC_Subscription` instead of an order but that caused different _type_ related issues. Namely, the [`from_payment_request `](https://github.com/Automattic/woocommerce-payments/blob/add/subscription_change_method/includes/class-payment-information.php#L199) expects an `WC_Order` but because in the testing environment the `WC_Subscription` doesn't extend the `WC_Order` that fails. 

I then tried to change the `WC_Subscription` mock class to extend a `WC_Order`, but that caused much broader issues.

If anyone has some ideas how we can keep these argument types, then let me know. 😅

Reply:
I think there is another case of this with [WC_Payments_Invoice_Service::get_subscription_invoice_id](https://github.com/Automattic/woocommerce-payments/blob/7c68a6f407eb8fe0c1497cec6856e85225a39d0c/includes/subscriptions/class-wc-payments-invoice-service.php#L263)

It looks like the [WC_Payment_Gateway_WCPay_Subscriptions_Process_Payment_Test](https://github.com/Automattic/woocommerce-payments/blob/develop/tests/unit/test-class-wc-payment-gateway-wcpay-subscriptions-process-payment.php#L170) tests trigger `maybe_record_first_invoice_payment` which attempts to get the invoice id from the subscription.

I've removed that type hint to get around this as well. I'm wondering if this is the preferred solution, or if we should open an issue for this and see if we can approach how we handle subscriptions in tests differently. What do you think?

Reply:
About these type hints (drive by comment!) - if the function supports different types for params, let's make sure the docs are correct, and ideally type hint as well.